contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -4029,3 +4035,309 @@ exec_stmt_fulltextindex(PLtsql_execstate *estate, PLtsql_stmt_fulltextindex *stm
+		List 			*arg = stmt->args;
+		bool 			isnull;
+		Oid			valtype;
+		int32			valtypmod;
+		Datum			tsql_type_datum;
+		char			*tsql_typename = NULL;
+		Datum			*values = NULL;
+		ArrayType		*arr_value = NULL;
+		Oid			sql_variant_oid;
+		tsql_compare_context 	cxt;
+		Operator		operator;
+		Oid			oprcode;
+		int 			nargs = list_length(arg);
+		MemoryContext		cur_ctxt = CurrentMemoryContext;
+		LOCAL_FCINFO(fcinfo, 1);
+
+		/* check if given name is exceeding the allowed limit */
+		if(strlen(partition_function_name) > 128)
+		{
+			ereport(ERROR, 
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("The identifier that starts with '%.128s' is too long. Maximum length is 128.", partition_function_name)));
+		}
+
+		/* check if there is existing partition function with the given name */
+		if(partition_function_exists(partition_function_name))
+		{
+			ereport(ERROR, 
+				(errcode(ERRCODE_DUPLICATE_FUNCTION),
+					errmsg("There is already an object named '%s' in the database.", partition_function_name)));
+		}
+
+		/* get the tsql typename from the input type */
+		InitFunctionCallInfoData(*fcinfo, NULL, 0, InvalidOid, NULL, NULL);
+		fcinfo->args[0].value = ObjectIdGetDatum(typ->typoid);
+		fcinfo->args[0].isnull = false;
+		tsql_type_datum = (*common_utility_plugin_ptr->translate_pg_type_to_tsql) (fcinfo);
+
+		if (tsql_type_datum)
+		{
+			tsql_typename = text_to_cstring(DatumGetTextPP(tsql_type_datum));
+		}
+
+		/*
+		 * check if datatype is supported or not, if tsql_typename is NULL
+		 * then it implies that typename corresponds to UDT
+		 */
+		if(!tsql_typename || (strcmp(tsql_typename, "text") == 0) || (strcmp(tsql_typename, "ntext") == 0)
+			|| (strcmp(tsql_typename, "image") == 0) || (strcmp(tsql_typename, "xml") == 0) || (strcmp(tsql_typename, "geometry") == 0)
+			|| (strcmp(tsql_typename, "geography") == 0) || (strcmp(tsql_typename, "timestamp") == 0) )
+		{
+			ereport(ERROR, 
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("The type '%s' is not valid for this operation.", typ->typname)));
+		}
+
+		/* check if the given number of boundaries are exceeding allowed limit */
+		if(nargs >= 15000)
+		{
+			ereport(ERROR, 
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					errmsg("CREATE/ALTER partition function failed as only a maximum of 15000 partitions can be created.")));
+		}
+
+		values = palloc0(nargs * sizeof(Datum));
+
+		for (int i = 0; i < nargs; i++)
+		{
+			Datum val;
+
+			/* evaluate the value from the expr */
+			val = exec_eval_expr(estate, list_nth(arg, i), &isnull, &valtype, &valtypmod);
+
+			/* 
+			 * implicitly convert range values to specified parameter type
+			 * and raise error with ordinal position if conversion fails
+			 */
+			PG_TRY();
+			{
+				values[i] = exec_cast_value(estate, val, &isnull,
+								valtype, valtypmod,
+								typ->typoid, typ->atttypmod);
+			}
+			PG_CATCH();
+			{
+				MemoryContext err_ctx;
+				err_ctx = MemoryContextSwitchTo(cur_ctxt);
```

Comment:
why memory context changes are required?

Reply:
We are changing the context to free the allocated memory in previous context.